### PR TITLE
apps/req: Modified key generation to use the type names

### PR DESCRIPTION
**Pull request by gotthardp**
This is to enable the key generation in external providers.

Particularly:
- added calls to `EVP_PKEY_CTX_new_from_pkey` and `EVP_PKEY_CTX_new_from_name`
- moved key_len checks to `set_keygen_ctx` to simplify the implementation

### CODE REVIEW

File: apps/req.c [hunk 1]
```diff
@@ -1515,61 +1478,34 @@ static int join(char buf[], size_t buf_size, const char *name,
...
-                *pkeylen = keylen;
-            } else {
-                keylen = *pkeylen;
-            }
-        } else if (p != NULL) {
-            paramfile = p + 1;
-        }
+            else
+                paramfile = p + 1;
+        } else
```
Comment (levitte) [inline apps/req.c:1507]: Nit: our coding style requires that if one branch of an `if` statement is wrapped with braces, then so must the other branch.

**Review by levitte — CHANGES_REQUESTED** [review]

File: apps/req.c [hunk 2]
```diff
@@ -1597,38 +1534,17 @@ static EVP_PKEY_CTX *set_keygen_ctx(const char *gstr,
...
-        }
-        EVP_PKEY_asn1_get0_info(NULL, NULL, NULL, NULL, &anam, ameth);
-        *palgnam = OPENSSL_strdup(anam);
-#ifndef OPENSSL_NO_ENGINE
-        finish_engine(tmpeng);
-#endif
-    }
-
-    if (param != NULL) {
-        gctx = EVP_PKEY_CTX_new(param, keygen_engine);
```
Comment (t8m) [inline apps/req.c:1627]: This removal is wrong. If the keygen_engine opt is set it must use the old way - with `EVP_PKEY_CTX_new` or `EVP_PKEY_CTX_new_id` based on whether we have param or not.

**Review by t8m — CHANGES_REQUESTED** [review]

File: apps/req.c (hunk 2 above)
Comment (gotthardp) [reply apps/req.c:1627]: Fixed, thanks!

File: apps/req.c [hunk 3]
```diff
@@ -1515,60 +1475,34 @@ static int join(char buf[], size_t buf_size, const char *name,
...
+        if (p != NULL) {
+            *palgnam = OPENSSL_strndup(gstr, p - gstr);
 
-        EVP_PKEY_asn1_get0_info(NULL, pkey_type, NULL, NULL, NULL, ameth);
-#ifndef OPENSSL_NO_ENGINE
-        finish_engine(tmpeng);
-#endif
-        if (*pkey_type == EVP_PKEY_RSA) {
-            if (p != NULL) {
+            if (!strcasecmp(*palgnam, "RSA") || !strcasecmp(*palgnam, "RSA-PSS"))
```
Comment (t8m) [inline apps/req.c:1500]: Use `== 0` please.

File: apps/req.c [hunk 4]
```diff
@@ -1597,38 +1531,27 @@ static EVP_PKEY_CTX *set_keygen_ctx(const char *gstr,
...
 
-    if (param != NULL) {
-        gctx = EVP_PKEY_CTX_new(param, keygen_engine);
-        *pkeylen = EVP_PKEY_bits(param);
         EVP_PKEY_free(param);
     } else {
-        gctx = EVP_PKEY_CTX_new_id(*pkey_type, keygen_engine);
+        int pkey_id;
+
+        pkey_id = get_legacy_pkey_id(app_get0_libctx(), *palgnam, keygen_engine);
```
Comment (t8m) [inline apps/req.c:1550]: This should be IMO done only if keygen_engine != NULL.

File: apps/req.c [hunk 5]
```diff
@@ -1641,7 +1564,22 @@ static EVP_PKEY_CTX *set_keygen_ctx(const char *gstr,
         EVP_PKEY_CTX_free(gctx);
         return NULL;
     }
-    if ((*pkey_type == EVP_PKEY_RSA) && (keylen != -1)) {
+    if ((EVP_PKEY_CTX_is_a(gctx, "RSA") || EVP_PKEY_CTX_is_a(gctx, "RSA-PSS"))
```
Comment (t8m) [inline apps/req.c:1567]: In the original code there was also handling of DSA for MIN_KEY_LENGTH and OPENSSL_DSA_MAX_MODULUS_BITS, can you please put it back?

**Review by t8m — CHANGES_REQUESTED** [review]
A few more minor changes needed, but good work!

File: apps/req.c (hunk 4 above)
Comment (gotthardp) [reply apps/req.c:1550]: I can do that, but please note that `genpkey.c`:`init_gen_str` does exactly this. The `get_legacy_pkey_id` feels ready for the situation when keygen_engine == NULL.

File: apps/req.c (hunk 5 above)
Comment (gotthardp) [reply apps/req.c:1567]: Yeah, it felt like a dead code, but it isn't. Sorry for that.

File: apps/req.c (hunk 4 above)
Comment (t8m) [reply apps/req.c:1556]: Ah ok then.

**Review by t8m — APPROVED** [review]

**Comment by gotthardp** [comment]
@levitte Please take another look.

**Review by t8m — APPROVED** [review]

**Comment by t8m** [comment]
Ping for second review

**Comment by gotthardp** [comment]
Replaced by #15400.

**Pull request closed without merging**